### PR TITLE
Curand switch

### PR DESCRIPTION
**Pull request by Robadob**
- [x] Move curand behind a typedef so it can be swapped at compile time (and forward the same to RTC builds)
- [x] Run the test suite for the two new curand options
- [x] Add a CMake option to swap curand
- [x] Change the default to Philox (#362 ?)
* ~~Add philox specific double/quad generation methods?~~ (#938)

~~Also fixed the warnings in test suite which @MILeach said they'd do weeks ago.~~ (Removed in rebase, as It had been resolved elsewhere)
Also added `USE_GLM` status to the RTC short reference as this had previously been missed.

Not assigning this to myself, someone else may pick it up to finish before me.

### CODE REVIEW

**Comment by Robadob** [comment]
Think I need to make sure the macro is passed to RTC builds too, forgot about that.

**Comment by ptheywood** [comment]
I'm not super keen on adding more library macros, which will make installing flamegpu system wide / providing pre-compiled binaries less flexible, but the required templating will probably be more painful than not. 

Though potentially as there are only 3 accepted curand types this could just be explicitly instantiated, with the macro required / used at model compilation instead?

With the philox 128byte versions, I'd probably rather make the api methods available for the other generators too so they models do not need specialising based on the value of the compile time macro (i.e. if someone needs 4 floats at once, they can just use the float4 version and philox will just run faster, where as the other generators will just be called 4 times. This will probably eat a few more registers, but shouldn't be too expensive)

**Comment by Robadob** [comment]
> I'm not super keen on adding more library macros, which will make installing flamegpu system wide / providing pre-compiled binaries less flexible, but the required templating will probably be more painful than not.


This is more a case of a hyper-advanced setting where we choose a default and don't expect anyone to ever have a reason to change it.

> Though potentially as there are only 3 accepted curand types this could just be explicitly instantiated, with the macro required / used at model compilation instead?

This just doesn't feel viable, same way we haven't included agent out state as a template arg to DeviceAPI.

> With the philox 128byte versions, I'd probably rather make the api methods available for the other generators too so they models do not need specialising based on the value of the compile time macro (i.e. if someone needs 4 floats at once, they can just use the float4 version and philox will just run faster, where as the other generators will just be called 4 times. This will probably eat a few more registers, but shouldn't be too expensive)

This seems fair.

**Comment by Robadob** [comment]
Now getting Link errors when I try to build with Philox. They don't make much sense, and I cba to dig into it right now. Pete says build's fine for him on Linux.

**Comment by Robadob** [comment]
Have now built and ran test suite for each of the 3 curand engine's consecutively. All worked as expected.

When running pyflamegpu test suite, it crashes with a fatal error (see end of this comment). However, that error is also currently present in the `master` pyflamegpu (windows/release) suite too. Hence, I consider this PR ready.


*This error only occurs running full suite, the individual `test_logging_exceptions.py` set of tests run fine in isolation.*
```
io\test_logging_exceptions.py ..Fatal Python error: Aborted

Current thread 0x00007aa8 (most recent call first):
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\capture.py", line 780 in pytest_runtest_setup
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_callers.py", line 34 in _multicall
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 259 in <lambda>
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 338 in from_call
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 258 in call_runtest_hook
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 219 in call_and_report
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 124 in runtestprotocol
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\runner.py", line 111 in pytest_runtest_protocol
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\main.py", line 347 in pytest_runtestloop
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\main.py", line 322 in _main
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\main.py", line 268 in wrap_session
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\main.py", line 315 in pytest_cmdline_main
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\config\__init__.py", line 164 in main
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\lib\site-packages\_pytest\config\__init__.py", line 187 in console_main
  File "C:\Users\Robadob\fgpu2\build2\lib\Release\python\venv\Scripts\py.test.exe\__main__.py", line 7 in <module>
  File "C:\ProgramData\Miniconda3\lib\runpy.py", line 87 in _run_code
  File "C:\ProgramData\Miniconda3\lib\runpy.py", line 197 in _run_module_as_main
Warning: Input file 'test.json' refers to second input file 'invalid', this will not be loaded.
RTC Initialisation Processing time: 0.000000 s
Warning: Input file 'test.json' refers to second input file 'invalid', this will not be loaded.
Step 1 Processing time: 0.000000 s
Warning: Input file 'test.xml' refers to second input file 'invalid', this will not be loaded.
RTC Initialisation Processing time: 0.000000 s
Warning: Input file 'test.xml' refers to second input file 'invalid', this will not be loaded.
Step 1 Processing time: 0.000000 s

(venv) (base) C:\Users\Robadob\fgpu2\tests\swig\python>
```

**Comment by ptheywood** [comment]
I've opened #938 to move the philox 128 bytes optimisation to a future API extension.

**Comment by ptheywood** [comment]
I've reproduced the vis  pyfalmegpu test suite failure in a release build under linux using current master (a76da8ca75a543513f5ea6c108a52da6f0c70afc)

```
../tests/swig/python/io/test_logging_exceptions.py ...Fatal Python error: Aborted
```
Will promote that to a separate issue.

File: include/flamegpu/util/detail/curand.cuh [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef INCLUDE_FLAMEGPU_UTIL_DETAIL_CURAND_CUH_
+#define INCLUDE_FLAMEGPU_UTIL_DETAIL_CURAND_CUH_
+
+/**
+ * This header exists to allow a convenient way to switch between curand implementations
+ */
+
+#include <curand_kernel.h>
+
```
[the remark this answers is not in the document]
Comment (Robadob) [reply include/flamegpu/util/detail/curand.cuh:9]: These are reasonable, should be able to address tomorrow.

[the remark this answers is not in the document]
Comment (Robadob) [reply include/flamegpu/util/detail/curand.cuh:9]: Resolved now, because it's split across 3 commits, the 1st commit doesn't have this fix but eh.

**Pull request merged**